src/cpu.js

```diff
@@ -1163,12 +1251,165 @@ CPU.prototype.load_multiboot = function(buffer)
+        // enable protected mode
+        data8[i++] = 0x0F; // mov eax, cr0
+        data8[i++] = 0x20;
+        data8[i++] = 0xC0;
+        data8[i++] = 0x0C; // or al, 0x1
+        data8[i++] = 0x01;
+        data8[i++] = 0x0F; // mov cr0, eax
+        data8[i++] = 0x22;
+        data8[i++] = 0xC0;
+        data8[i++] = 0x66; // jmp 0x8:0xFFFFFFFF (placeholder to be replaced by physical address of trampoline)
+        data8[i++] = 0xEA;
+        dbg_assert(i == TRAMPOFFSET, "trampoline offset is not correct");
+        data8[i++] = 0xFF;
+        data8[i++] = 0xFF;
+        data8[i++] = 0xFF;
+        data8[i++] = 0xFF;
+        data8[i++] = 0x08;
+        data8[i++] = 0x00;
+        dbg_assert(i == GDTOFFSET, "GDTOFFSET is not correct");
+        // GDT
+        data8[i++] = GDTSIZE - 1; // GDT[0] is inaccesible, use it to store the GDT Descriptor
+        data8[i++] = 0;
+        data8[i++] = 0; data8[i++] = 0;
+        data8[i++] = 0; data8[i++] = 0; data8[i++] = 0; data8[i++] = 0;
+        // segment 0x08, Ring 0 code, base 0, limit 0xFFFF FFFF
+        data8[i++] = 0xFF;
+        data8[i++] = 0xFF;
+        data8[i++] = 0;
+        data8[i++] = 0;
+        data8[i++] = 0;
+        data8[i++] = 0x9A;
+        data8[i++] = 0xCF;
+        data8[i++] = 0;
+        // segment 0x10, Ring 0 data, base 0, limit 0xFFFF FFFF
+        data8[i++] = 0xFF;
+        data8[i++] = 0xFF;
+        data8[i++] = 0;
+        data8[i++] = 0;
+        data8[i++] = 0;
+        data8[i++] = 0x92;
+        data8[i++] = 0xCF;
+        data8[i++] = 0;
+        dbg_assert(i - GDTOFFSET == GDTSIZE, "GDTSIZE is not correct");
+        // now in protected mode, set the rest of the segments
+        data8[i++] = 0xB8; // mov ax, 0x10
+        data8[i++] = 0x10;
+        data8[i++] = 0x00;
+        data8[i++] = 0x8e; // mov ds, eax
+        data8[i++] = 0xd8;
+        data8[i++] = 0x8e; // mov es, eax
+        data8[i++] = 0xc0;
+        data8[i++] = 0x8e; // mov fs, eax
+        data8[i++] = 0xe0;
+        data8[i++] = 0x8e; // mov gs, eax
+        data8[i++] = 0xe8;
+        data8[i++] = 0x8e; // mov ss, eax
+        data8[i++] = 0xd0;
+        // trigger load and input entrypoint into eax
+        data8[i++] = 0xe5; // in 0xF4
+        data8[i++] = 0xF4;
+        // copy entrypoint address to ecx
+        data8[i++] = 0x89; // mv ecx, eax
+        data8[i++] = 0xc1;
+        // place magic value in eax
+        data8[i++] = 0xb8; // mov eax, 0x2badb002
+        data8[i++] = 0x02;
+        data8[i++] = 0xb0;
+        data8[i++] = 0xad;
+        data8[i++] = 0x2b;
+        // place multiboot header location in ebx
+        data8[i++] = 0xbb; // mov ebx, multiboot_info_addr
+        data8[i++] = multiboot_info_addr;
+        data8[i++] = multiboot_info_addr >> 8;
+        data8[i++] = multiboot_info_addr >> 16;
+        data8[i++] = multiboot_info_addr >> 24;
+        // jump to multiboot entrypoint
+        data8[i++] = 0xFF; // jmp rcx
+        data8[i++] = 0xE1;
```

Comment:
Since we're using a (somewhat ugly) io port hook anyway, wouldn't it be simpler to implement all of this in JavaScript in the hook?

It would also remove this extra work from nasmtests (i.e. load_multiboot could be implement as exactly this callback plus setting cs/eip, if that makes sense)

Reply:
That's a good idea! We need a small boot rom so that we can trigger the io port hook after bios does its thing, but it can be really short.